src/test/socket/bind/test_bind.rs

```diff
@@ -0,0 +1,588 @@
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let addr2 = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 11111u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_ANY.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    // if reverse, bind to ANY before LOOPBACK
+    let (addr1, addr2) = if reverse {
+        (addr2, addr1)
+    } else {
+        (addr1, addr2)
+    };
+
+    let args1 = BindArguments {
+        fd: fd1,
+        addr: Some(LibcSockAddr::In(addr1)),
+        addr_len: None,
+    };
+
+    let args2 = BindArguments {
+        fd: fd2,
+        addr: Some(LibcSockAddr::In(addr2)),
+        addr_len: None,
+    };
+
+    run_and_close_fds(&[fd1, fd2], || {
+        check_bind_call(&args1, false, None)?;
+        check_bind_call(&args2, true, Some(libc::EADDRINUSE))?;
+        Ok(())
+    })?;
+
+    Ok(())
+}
+
+// test binding to all interfaces with a port of 0
+fn test_unspecified_port(sock_type: libc::c_int, flag: libc::c_int) -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd >= 0);
+
+    let addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 0u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_ANY.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let args = BindArguments {
+        fd: fd,
+        addr: Some(LibcSockAddr::In(addr)),
+        addr_len: None,
+    };
+
+    run_and_close_fds(&[fd], || check_bind_call(&args, false, None))?;
+
+    Ok(())
+}
+
+// run the function and then close any given file descriptors, even if there was an error
+fn run_and_close_fds<F>(fds: &[libc::c_int], f: F) -> Result<(), String>
+where
+    F: Fn() -> Result<(), String>,
+{
+    let rv = f();
+
+    for fd in fds.iter() {
+        let fd = *fd;
+        let rv_close = unsafe { libc::close(fd) };
+        assert_eq!(rv_close, 0, "Could not close the fd");
+    }
+
+    rv
+}
+
+fn get_errno() -> i32 {
+    std::io::Error::last_os_error().raw_os_error().unwrap()
+}
+
+fn get_errno_message(errno: i32) -> String {
+    let cstr;
+    unsafe {
+        let error_ptr = libc::strerror(errno);
+        cstr = std::ffi::CStr::from_ptr(error_ptr)
+    }
+    cstr.to_string_lossy().into_owned()
+}
+
+fn check_bind_call(
+    args: &BindArguments,
+    should_error: bool,
+    expected_errno: Option<libc::c_int>,
+) -> Result<(), String> {
+    // get a pointer to the sockaddr and the size of the structure
+    // careful use of references here makes sure we don't copy memory, leading to stale pointers
+    let (addr_ptr, real_addr_len) = match &args.addr {
+        Some(addr) => match &addr {
+            &LibcSockAddr::In(sockaddr_in) => (
+                sockaddr_in as *const _ as *const libc::sockaddr,
```

Comment:
I'm having trouble following this. Could you walk me through it a bit? (Might be worth a comment, but OTOH might not if it's just a matter of me not knowing the language well enough yet)

Reply:
I added some comments here to hopefully make it clearer, and I replaced the `_` with an explicit type. But `bind()` requires the underlying address type (ex: `libc::sockaddr_in` or `libc::sockaddr_in6`) to be cast to a `libc::sockaddr`. So the `sockaddr_in as *const _` casts the reference of type `&libc::sockaddr_in` to a pointer `*const libc::sockaddr_in`, and the `as *const libc::sockaddr` switches the pointer type from `*const libc::sockaddr_in` to `*const libc::sockaddr`. Rust requires the two casts since you cannot cast a reference (`&sockaddr_in`) directly to a pointer of a different type (`*const libc::sockaddr`).